src/docs/user/ProActiveUserGuide.adoc

```diff
@@ -1763,6 +1710,234 @@ println outFile.text
+
+In our scenario, the number of available slots is immediately decremented to reserve a slot and enter the critical section.
+
+Releasing of the slot is performed at the end of the task using the https://www.activeeon.com/public_content/documentation/javadoc/dev/org/ow2/proactive/scheduler/synchronization/Synchronization.html#compute-java.lang.String-java.lang.String-java.lang.String-[compute] method.
+
+ * *Clean*: Delete the channel. Same code as the previous examples.
+
+Here is the output of the Example 3 workflow:
+```
+[3t0@trydev.activeeon.com;16:14:14] Channel 3 created.
+[3t1@trydev.activeeon.com;16:14:22] Waiting for available slot
+[3t4@trydev.activeeon.com;16:14:22] Waiting for available slot
+[3t3@trydev.activeeon.com;16:14:22] Waiting for available slot
+[3t5@trydev.activeeon.com;16:14:22] Waiting for available slot
+[3t1@trydev.activeeon.com;16:14:23] Enter critical section
+[3t4@trydev.activeeon.com;16:14:23] Enter critical section
+[3t4@trydev.activeeon.com;16:14:33] Release slot
+[3t1@trydev.activeeon.com;16:14:33] Release slot
+[3t3@trydev.activeeon.com;16:14:33] Enter critical section
+[3t5@trydev.activeeon.com;16:14:34] Enter critical section
+[3t6@trydev.activeeon.com;16:14:38] Waiting for available slot
+[3t3@trydev.activeeon.com;16:14:43] Release slot
+[3t5@trydev.activeeon.com;16:14:44] Release slot
+[3t6@trydev.activeeon.com;16:14:44] Enter critical section
+[3t6@trydev.activeeon.com;16:14:54] Release slot
+[3t2@trydev.activeeon.com;16:14:59] Channel 3 deleted.
+```
+
+We can see that no more than two replicated tasks enter the critical section at the same time, and that releasing slots allows new tasks to enter the critical section.
+
+The Complete Workflow example can be downloaded link:examples/CriticalSection.xml[here].
+
+CAUTION: Implementing the Critical Section example using <<_selection,selection scripts>> rather than explicit wait methods is more *complex*.
+The reason behind is that selection scripts are meant to be *stateless*.
+The same selection script is executed on all ProActive Nodes by the Scheduler before deciding if a Node can be used to execute a single Task.
+In our use case, we must both decide *atomically* if a slot is available (_stateless_) and reserve a slot (_stateful_). This decision will be performed *simultaneously* on all available ProActive Nodes.
+An example implementation can be downloaded link:examples/CriticalSectionWithSelectionScripts.xml[here]. It makes use of a `reserved_slots` binding containing a set of `task ids`.
+
+==== Conclusion
+
+Task Synchronization API can be used for multiple dynamic synchronization patterns such as:
+
+ * *Cross-workflow* synchronization
+
+ * *Critical Sections*
+
+ * *Barriers* (Wait for several parallel tasks to reach a certain point)
+
+ * *Producer / Consumer*
+
+A complete example of Producer / Consumer, _ServiceDeployerConsumer_ can be downloaded link:examples/ServiceDeployerConsumer.xml[here].
+
```

Comment:
I think that what might be missing from the Synchronization API section is maybe a note on how errors in Groovy lambdas are handled, how the user can spot them and fix them.

Reply:
I added one line in the API Usage section